packages/query-core/src/queryClient.ts

```diff
@@ -129,12 +130,17 @@ export class QueryClient {
     TData = TQueryFnData,
     TQueryKey extends QueryKey = QueryKey,
   >(
-    options: FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    options: EnsureQueryDataOptions<TQueryFnData, TError, TData, TQueryKey>,
   ): Promise<TData> {
-    const cachedData = this.getQueryData<TData>(options.queryKey)
+    const defaultedOptions = this.defaultQueryOptions(options)
+    const query = this.#queryCache.build(this, defaultedOptions)
+
+    if (query.state.data && query.isStaleByTime(defaultedOptions.staleTime)) {
+      query.fetch(defaultedOptions)
+    }
 
-    return cachedData !== undefined
-      ? Promise.resolve(cachedData)
+    return query.state.data !== undefined
+      ? Promise.resolve(query.state.data)
       : this.fetchQuery(options)
   }
```

Comment:
I think we should generally implement this a bit differently, flow-wise:

- first, look up the `cachedData`
- if it's undefined
  - return fetch
- else 
  - check if that data is stale
  - if it is - trigger a background refetch
  - in any case, return it

```
const cachedData = this.getQueryData<TData>(options.queryKey)

if (cachedData === undefined) {
  return this.fetchQuery(options)
}

const defaultedOptions = this.defaultQueryOptions(options)
if (query.isStaleByTime(defaultedOptions.staleTime)) {
  void this.prefetchQuery(defaultedOptions)
}

return cachedData
```

calling `prefetchQuery` instead of `query.fetch` is important because:

- it catches errors that happen during fetching, which is what we want for a background refetch
- it goes through `fetchQuery`, which turns off retries, which is also what we want:

https://github.com/TanStack/query/blob/bdd2b552d3647242e45c85f4c765fd293862d6d5/packages/query-core/src/queryClient.ts#L329-L331